dissect-image: open dissected and decrypted partitions

yuwata opened the pull request:
If multiple services with the same encrypted image are simultaneously
starting, one may deactivate the dm device while others using it.
To prevent such the situation, let's open the dm device
for the decrypted volume.

Fixes #24617.

poettering commented on src/shared/dissect-image.c:
```diff
@@ -2061,6 +2070,33 @@ static int verity_partition(
+
+        rw = p->rw && !(flags & DISSECT_IMAGE_MOUNT_READ_ONLY);
+
+        fd = open(p->decrypted_node, O_NONBLOCK | O_CLOEXEC | O_NOCTTY | (rw ? O_RDWR : O_RDONLY));
```
we can always open this in O_RDONLY mode. We just want to pin, not use for writes. And if we mount via /proc/self/fd/ then the kernel will only look at the inode we pin, and not what kind of fd we have...

yuwata replied:
Is it possible to use O_PATH?

poettering replied:
I dont think so. I am pretty sure if you use O_PATH the bsd locks would end up being unaffected by the bsd locks on the non-O_PATH fds...

yuwata replied:
OK. Thanks.

poettering commented on src/shared/dissect-image.h:
```diff
@@ -26,6 +26,7 @@ struct DissectedPartition {
         char *fstype;
         char *node;
         char *label;
+        int decrypted_node_fd;
```
hmm, instead of doing this only if decryption is in place, I'd always pin the device here once we know which one to mount. i.e. call this `mount_node_fd` or so, then always open the block device one we figured it out both if encrypted or if not, and then always mount through /proc/self/fd/.

It think that's much prettier since it adds less special casing to the encryption code path

poettering commented on src/shared/dissect-image.c:
```diff
@@ -228,6 +229,31 @@ static int make_partition_devname(
+
+        assert(p);
+
+        if (!p->decrypted_node && p->mount_node_fd >= 0)
```
why the `!p->decrypted_node` check?

yuwata replied:
First, `dissect_image()` opens all partitions and store their fd into `p->mount_node_fd`.
Then, `dissected_image_decrypt()` opens all decrypted partitions and replaces existing fds.
Hence, `p->mount_node_fd` may be for `p->node`, instead of `p->decrypted_node`.

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1268,12 +1314,13 @@ static int mount_partition(
         assert(m);
         assert(where);
 
+        if (!m->found || m->mount_node_fd < 0)
```
the left-hand check is redundant, no? the right-hand check already should be enough?

poettering commented on src/shared/dissect-image.c:
```diff
@@ -228,6 +229,31 @@ static int make_partition_devname(
+        safe_close(p->mount_node_fd);
+        p->mount_node_fd = TAKE_FD(fd);
+
+        return 0;
```
i think it would make sense to to compare the BLKGETDISKSEQ value of these partition devices with the one of the "whole" device here, to ensure we don't end up with wrong devices.

And beforehand we should check that what we just opened is actually a block device. 

And this call should probably not even try to open() if the "whole" node is not actually a block device itself.

yuwata replied:
OK.
(But, you said in my other PRs, `dissect_image()` should work for non device fd...)

yuwata commented:
@poettering and @bluca Rebased on top of my recent cleanup commits. PTAL!

bluca commented on src/shared/dissect-image.c:
```diff
@@ -2068,59 +2068,68 @@ static int verity_partition(
+                goto success;
+
+        try_again:
+                /* Device is being removed by another process. Let's wati for a while. */
```
/wati/wait/

bluca commented on src/shared/dissect-image.c:
```diff
@@ -2069,7 +2069,16 @@ static int verity_partition(
                 _cleanup_(sym_crypt_freep) struct crypt_device *existing_cd = NULL;
+                _cleanup_close_ int fd = -1;
 
+                /* First, check the device is already exist. */
```
"check if the device already exists"

bluca reviewed (commented):
commit message mentions a lock is taken, but where does that happen?

yuwata commented:
> commit message mentions a lock is taken, but where does that happen?

Sorry, I forgot to update the commit message. Now about the lock is dropped from the commit message.

At least in this PR, the partitions are opened but not locked, as udevd locks only whole (native, i.e., non-DM) block devices. So, I cannot find any reasons we need to lock dissected partitions. Also, the whole block device to be dissected may be already locked through `loop_device_open()` or `loop_device_make()`. It may be better to call `loop_device_flock()` in `dissect_loop_device()`. But, I'd like to do that in another PR if necessary.

yuwata commented:
@bluca Thank you for reviewing this PR.
@poettering Could you re-review this if you'd like?

keszybz commented on src/shared/dissect-image.c:
```diff
@@ -264,6 +265,47 @@ static int make_partition_devname(
         return asprintf(ret, "%s%s%i", whole_devname, need_p ? "p" : "", nr);
 }
 
+static int open_partition(const char *node, bool is_partition, LoopDevice *loop) {
```
I think the signature should be:
```C
static int open_partition(const LoopDevice *loop, const char *node, bool is_partition)
```

yuwata replied:
The main target is `node`. And `loop` is used for checking the opened file descriptor. I will add the `const` qualifier for `loop`.

yuwata commented:
@keszybz Thank you for your review. Addressed the above comments, except for the one for the order of the arguments in `open_partition()`. Hopefully, that should not be a big matter. Upgrading the green label.
